### ISSUE TITLE
Regex.split/3 behaves differently on Elixir 1.6.1 compared to 1.5.1

### ISSUE THREAD

**Issue by rradz** [elixir-lang/elixir#7385]
Regex.split/3 function changed behaviour on migration from 1.5.1 to 1.6.1, effectively breaking Inflex library. I checked the changelog for 1.6.0 and 1.6.1, no changes to the function where reported, so I assume this might be a bug.

### Environment
OTP 20, Elixir 1.5.1 & 1.6.1
Mac OS X/Linux

### Current behavior

```elixir
Regex.split(~r/(?:^|[-_])|(?=[A-Z])/, "lower")
["", "lower"]
```

### Expected behavior
```elixir
Regex.split(~r/(?:^|[-_])|(?=[A-Z])/, "lower")
["lower"]
```

**Comment by josevalim** [comment]
@rradz it is a bug fix. Replace should be equivalent to a split + join.

This is how Regex.replace has always worked:

    iex> Regex.replace(~r/(?:^|[-_])|(?=[A-Z])/, "lower", "a")
    alower

However, the previous split behaviour would not given the same result on a join. So we have addressed this bug and inconsistency. I will update the changelog. Thank you!

**Comment by lukaszsamson** [comment]
Shouldn't that be in Breaking changes also? It did break my code also :)

**Comment by josevalim** [comment]
In theory, fixing any bug can be considered backwards incompatible behaviour. That's why we have outlined what to expect from our releases back when we released 1.0: https://elixir-lang.org/blog/2014/09/18/elixir-v1-0-0-released/. I will also port those expectations to our docs.

**Comment by ssboisen** [comment]
Just FYI this breaking change also caught us off guard via a dependency on [Inflex](https://github.com/nurugger07/inflex). I think going forward it would be beneficial to also mention bug fixes that are known breaking changes to public api in a separate breaking changes list.

Thanks :)
